### ISSUE TITLE
Error: Anomaly "Uncaught exception Not_found." (explain_ill_formed_rec_body)

### ISSUE THREAD

**Issue by mdempsky** [coq/coq#11829]
#### Description of the problem

Feeding the input below to coqtop causes it to print:

    Error: Anomaly "Uncaught exception Not_found."
    Please report at http://coq.inria.fr/bugs/.

Immediately before inputting the second "Qed.", it prints "No more subgoals."

#### Coq Version

```coq
$ coqtop --version
The Coq Proof Assistant, version 8.11.0 (March 2020)
compiled on Mar 5 2020 8:13:48 with OCaml 4.05.0
```

```coq
Require Import List.
Import ListNotations.

Inductive Tree : Set :=
| Leaf : nat -> Tree
| Fork : Tree -> Tree -> Tree
.

Definition list1 T := { l : list T | l <> [] }.

Definition compose {A B C} (f : B -> C) (g : A -> B) x := f (g x).

Definition concatMap {A B} (f : A -> list B) := compose (@concat _) (map f).

Program Fixpoint prefixes x t : list1 Tree :=
  match t with
  | Leaf y => [Fork (Leaf x) t]
  | Fork u v => [Fork (Leaf x) t] ++ map (fun u' => Fork u' v) (prefixes x u)
  end.

Program Fixpoint trees (xs : list1 nat) : list1 Tree :=
  match xs with
  | [] => _
  | [x] => [Leaf x]
  | x::xs => concatMap (prefixes x) (trees (_ xs))
  end.
Next Obligation.
  destruct xs.
  cbn in Heq_xs.
  subst.
  exfalso.
  apply n.
  reflexivity.
Qed.
Next Obligation.
  unfold concatMap.
  unfold compose.
  destruct x0.
  - exfalso. apply n0. reflexivity.
  - cbn.
    intros H.
    destruct (prefixes x t).
    cbn in H.
    destruct x1.
    + intuition.
    + cbn in H. discriminate H.
Qed.
```

**Comment by ejgallego** [comment]
Seems fixed in `master`, or at least if doesn't produce an anomaly but an error:

```coq
Error:
Recursive definition of trees is ill-formed.
In environment
trees : list1 nat -> list1 Tree
xs : list1 nat
x : nat
xs0 : list nat
n : nat
l : list nat
Heq_xs : x :: n :: l = proj1_sig xs
Recursive call to trees has principal argument equal to

```

**Comment by ejgallego** [comment]
It seems the call to `Pretyping.search_guard` happens in the wrong env, maybe some of the envs tricks played by the STM?

**Comment by herbelin** [comment]
Partial debugging:

The problem is related to the error being printed in the wrong `Nametab` trying to refer to `trees_obligation_5` which it did not find.

In both coqc and coqtop, the text (`Pp.t`) of the error is actually computed twice:
- once in `exn_on` (from within `Stm.State.define`), but it is just sent to `feedback`
- once later by `read_and_execute`

The first one is fine in both `coqc` and `coqtop`. The second is in the wrong state for `coqtop` but not for `coqc`. The difference is in the `interactive` flag in `interp_vernac` which is true in `coqtop` and which causes an `Stm.edit_at ~doc:state.doc state.sid` in `coqtop` which does not occur in `coqc`.

Pending question: why is the error computed twice?

**Comment by ejgallego** [comment]
> Pending question: why is the error computed twice?

Maybe related to https://github.com/coq/coq/issues/5479

**Comment by Alizter** [comment]
Still an issue in 8.15+alpha.

**Comment by SkySkimmer** [comment]
Fixed since 8.16
